test/Storage.js

```diff
@@ -0,0 +1,160 @@
+
+        expect(construction.options.port).to.be.equal(80);
+      });
+    });
+
+    describe('when constructed with https protocol and no port', () => {
+      it('should use https port (443)', () => {
+        const httpsSettings = { hostname: 'data.knot.cloud', id: '5770d58c-a25c-40bc-ac1d-8613002be539', token: 'cff126c40a47438a33e8451c87a4199fddbdbb40', protocol: 'https' };
+        const construction = new Storage(httpsSettings);
+
+        expect(construction.options.port).to.be.equal(443);
+      });
+    });
+
+    describe('when constructed without hostname', () => {
+      it('should throw an exception', () => {
+        expect(() => new Storage({ port: 80, protocol: 'http' }))
+          .to.throw();
+      });
+    });
+
+    describe('when constructed without uuid and token', () => {
+      it('should throw an exception', () => {
+        expect(() => new Storage({ hostname: 'data.knot.cloud', port: 80, protocol: 'http' }))
+          .to.throw();
+      });
+    });
+  });
+
+  describe('listData', () => {
+    let storage, query;
+
+    describe('with empty query', () => {
+      beforeEach(() => {
+        storage = new Storage(settings);
+        nock(url)
+          .get('/data')
+          .reply(200, dataList);
+      });
+
+      it('should returns an array', async () => {
+        const response = await storage.listData();
+        expect(response).to.be.an('array');
+      });
+
+      it('should returns an array of data objects', async () => {
+        const response = await storage.listData();
+        expect(response).to.deep.equal(dataList);
+      });
+    });
+
+    describe('with date interval query', () => {
+      let dataBetweenTimeInterval;
+
+      beforeEach(() => {
+        query = { startDate: '2019-03-18T14:42:03.192Z', finishDate: '2019-03-18T16:17:45.192Z'};
+        dataBetweenTimeInterval = dataList.slice(1, 4);
+        storage = new Storage(settings);
+        nock(url)
+          .get(`/data`)
+          .query(query)
+          .reply(200, dataBetweenTimeInterval);
+        });
+
+      it('should returns an array', async () => {
+        const response = await storage.listData(query);
+        expect(response).to.be.an('array');
+      });
+
+      it('should returns an array of data objects between time interval', async () => {
+        const response = await storage.listData(query);
+        expect(response).to.deep.equal(dataBetweenTimeInterval);
+      });
+    });
+
+    describe('with take and skip query', () => {
+      let skippedData;
+
+      beforeEach(() => {
+        query = { skip: 3, take: 2 };
+        skippedData = dataList.slice(query.skip, dataList.length);
+        storage = new Storage(settings);
+        nock(url)
+          .get('/data')
+          .query(query)
+          .reply(200, skippedData);
+      });
+
+      it('should returns an array', async () => {
+        const response = await storage.listData(query);
+        expect(response).to.be.an('array');
```

Comment:
Here you are validating the response that you set in the mock and this is already being made by the previous test. The result of passing a query is actually determined by the server, which isn't being tested here. What should be being tested is the relation between the library code and another external component (e.g. axios) or what changes in the output based on the input.

For instance, you could test that when the user passes a query as parameter, the library encodes the query correctly in a query string. Or when the user specifies `skip` and `take`, these arguments are encoded in the query string.

Reply:
Nice, I got it. I'm already validating if the request is made correctly since the `nock` library will wait for the exactly match of the request. So, considering that I've defined this request:

```
        nock(url)
          .get(`/data`)
          .query(query)
          .reply(200, dataBetweenTimeInterval);
        });
```

If any request did by `storage.listData(query)` doesn't encode the argument into a valid query string, it will throw an error. Should I do only one test to validate if a request with query is ok and remove the other ones?

Reply:
Done. I've added some tests to validate the interaction between the lib and axios (with `nock`).